IO/BeagleBone/CANBusBBB.cs

```diff
@@ -1,169 +1,301 @@
+            }
+        }
+
+        /// <summary> Reads a CAN frame asynchronously, returning the payload and the ID of the received CAN frame. </summary>
+        /// <returns> A task that will return a tuple, with the first element being the ID of the received CAN frame and the second being the payload </returns>
+        public Task<Tuple<uint, byte[]>> ReadAsync()
+        {
+            return Task.Run(() =>
+            {
+                if (Extended)
+                {
+                    ExtendedCANFrame Frame = new ExtendedCANFrame();
+                    read(Socket, ref Frame, Marshal.SizeOf(Frame));
+                    byte[] Payload = new byte[Frame.DataLength];
+                    unsafe
+                    {
+                        for (int i = 0; i < Frame.DataLength; i++) { Payload[i] = Frame.Data[i]; }
+                    }
+                    return new Tuple<uint, byte[]>(Frame.CANID, Payload);
+                }
+                else
+                {
+                    CANFrame Frame = new CANFrame();
+                    read(Socket, ref Frame, Marshal.SizeOf(Frame));
+                    byte[] Payload = new byte[Frame.DataLength];
+                    unsafe
+                    {
+                        for (int i = 0; i < Frame.DataLength; i++) { Payload[i] = Frame.Data[i]; }
+                    }
+                    return new Tuple<uint, byte[]>(Frame.CANID, Payload);
+                }
+            });
+        }
+
+        /// <summary> Converts a DLC into a proper multiple of 4 for CAN transmission </summary>
+        /// <returns> The multipole of 4 </returns>
+        /// <param name="CanDLC"> Can dlc. </param>
+        private byte DLCToLength(byte CanDLC)
+        {
+            byte[] DLC2Len = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64 };
+            return DLC2Len[CanDLC & 0x0F];
+        }
+
+        /// <summary> Converts a length to a CAN data length for use with DLCToLength. </summary>
+        /// <returns> The DLC. </returns>
+        /// <param name="CanLen"> Can length. </param>
+        private byte LengthToDLC(byte CanLen)
+        {
+            byte[] Len2DLC =
+            {
+                0, 1, 2, 3, 4, 5, 6, 7, 8,       /* 0 - 8 */
+                    9, 9, 9, 9,             /* 9 - 12 */
+                    10, 10, 10, 10,             /* 13 - 16 */
+                    11, 11, 11, 11,             /* 17 - 20 */
+                    12, 12, 12, 12,             /* 21 - 24 */
+                    13, 13, 13, 13, 13, 13, 13, 13,     /* 25 - 32 */
+                    14, 14, 14, 14, 14, 14, 14, 14,     /* 33 - 40 */
+                    14, 14, 14, 14, 14, 14, 14, 14,     /* 41 - 48 */
+                    15, 15, 15, 15, 15, 15, 15, 15,     /* 49 - 56 */
+                15, 15, 15, 15, 15, 15, 15, 15 /* 57 - 64 */
+            };
+            if (CanLen > 64) { return 0xF; }
+            return Len2DLC[CanLen];
+        }
```

Comment:
These two method names still aren't clear. Below, you do `DLCToLength(LengthToDLC())`, which looks like a pointless conversion, even though it's not. Change the names to something more clear and descriptive.